karpenter-operator/controllers/aws/ec2_nodeclass.go

```diff
@@ -0,0 +1,376 @@
+					return []ctrl.Request{{NamespacedName: client.ObjectKey{Namespace: r.Namespace}}}
+				}
+				return nil
+			},
+		)).
+		Watches(&awskarpenterv1.EC2NodeClass{}, &handler.EnqueueRequestForObject{})
+
+	return bldr.Complete(r)
+}
+
+func (r *EC2NodeClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling", "req", req)
+
+	hcp, err := r.getHCP(ctx)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileCRDs(ctx, false); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	openshiftEC2NodeClass := &hyperkarpenterv1.OpenshiftEC2NodeClass{}
+	if err := r.guestClient.Get(ctx, req.NamespacedName, openshiftEC2NodeClass); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("openshiftEC2NodeClass not found, aborting reconcile", "name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get openshiftEC2NodeClass %q: %w", req.NamespacedName, err)
+	}
+
+	userDataSecret, err := r.getUserDataSecret(ctx, hcp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ec2NodeClass := &awskarpenterv1.EC2NodeClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      openshiftEC2NodeClass.Name,
+			Namespace: openshiftEC2NodeClass.Namespace,
+		},
+	}
+
+	if _, err := r.CreateOrUpdate(ctx, r.guestClient, ec2NodeClass, func() error {
+		return reconcileEC2NodeClass(ec2NodeClass, openshiftEC2NodeClass, hcp, userDataSecret)
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileStatus(ctx, openshiftEC2NodeClass); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileVAP(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcileCRDs reconcile the Karpenter CRDs, if onlyCreate is true it uses an only write non cached client.
+func (r *EC2NodeClassReconciler) reconcileCRDs(ctx context.Context, onlyCreate bool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	errs := []error{}
+	var op controllerutil.OperationResult
+	var err error
+	for _, crd := range []*apiextensionsv1.CustomResourceDefinition{
+		crdEC2NodeClass,
+		crdOpenshiftEC2NodeClass,
+	} {
+		if onlyCreate {
+			if err := r.guestClient.Create(ctx, crd); err != nil {
+				if !apierrors.IsAlreadyExists(err) {
+					errs = append(errs, err)
+				}
+			}
+		} else {
+			op, err = r.CreateOrUpdate(ctx, r.guestClient, crd, func() error {
+				return nil
+			})
+			if err != nil {
+				errs = append(errs, err)
+			}
+
+		}
+	}
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		return fmt.Errorf("failed to reconcile CRDs: %w", err)
+	}
+	log.Info("Reconciled CRDs", "op", op)
+
+	return nil
+}
+
+func reconcileEC2NodeClass(ec2NodeClass *awskarpenterv1.EC2NodeClass, openshiftEC2NodeClass *hyperkarpenterv1.OpenshiftEC2NodeClass, hcp *hyperv1.HostedControlPlane, userDataSecret *corev1.Secret) error {
```

Comment:
we need units for the mirroring reconciliation. Might be also nice to have fuzz testing

Reply:
updated the unit test